### PR TITLE
Add VERSION_QUALIFIER support for use by release manager

### PR DESCRIPTION
**Pull request by guyboertje**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -366,11 +372,9 @@ task generateLicenseReport(type: JavaExec) {
 
     classpath = project.files([jarFile])
     main = "org.logstash.dependencies.Main"
-    args \
-      licenseReportInputCSV, \
```
Comment (guyboertje) [inline build.gradle]: IntelliJ complained about the line continuations.

File: build.gradle [hunk 2]
```diff
@@ -409,9 +413,17 @@ bootstrap.dependsOn installTestGems
 runIntegrationTests.shouldRunAfter tasks.getByPath(":logstash-core:test")
 check.dependsOn runIntegrationTests
 
+String artifactsVersionApi = "https://artifacts-api.elastic.co/v1/versions/"
+def dlVersions = new JsonSlurper().parseText(artifactsVersionApi.toURL().text)
+// the version string can be either '7.0.0' or '7.0.0-alpha1', i.e. with the qualifier.
+// in the normal PR type builds it is plain '7.0.0'
+// in the build invoked by the release manager it is '7.0.0-alpha1' etc.
+// the artifacts-api will return JSON like this: `{"versions":["5.6.13-SNAPSHOT","6.4.3-SNAPSHOT","6.5.0-SNAPSHOT","6.6.0-SNAPSHOT","7.0.0-alpha1-SNAPSHOT"]}`
+String qualifiedVersion = dlVersions['versions'].grep(~/^${version}.*/)[0]
```
Comment (colinsurprenant) [inline build.gradle]: should this be validated?

File: logstash-core-plugin-api/logstash-core-plugin-api.gemspec [hunk 3]
```diff
@@ -28,7 +28,7 @@ Gem::Specification.new do |gem|
   gem.description   = %q{Logstash plugin API}
   gem.summary       = %q{Define the plugin API that the plugin need to follow.}
   gem.homepage      = "http://www.elastic.co/guide/en/logstash/current/index.html"
-  gem.license       = "Apache License (2.0)"
+  gem.license       = "Apache-2.0"
```
Comment (colinsurprenant) [inline logstash-core-plugin-api/logstash-core-plugin-api.gemspec]: this seems unrelated? is that new license string consistent across gemspecs?

File: rakelib/artifacts.rake [hunk 4]
```diff
@@ -1,7 +1,18 @@
...
-  PACKAGE_SUFFIX = SNAPSHOT_BUILD ? "-SNAPSHOT" : ""
+  VERSION_QUALIFIER = ENV["VERSION_QUALIFIER"]
+  if VERSION_QUALIFIER
+    if SNAPSHOT_BUILD
+      PACKAGE_SUFFIX = "-#{VERSION_QUALIFIER}-SNAPSHOT"
+    else
+      PACKAGE_SUFFIX = "-#{VERSION_QUALIFIER}"
+    end
+  elsif SNAPSHOT_BUILD
+      PACKAGE_SUFFIX = "-SNAPSHOT"
```
Comment (colinsurprenant) [inline rakelib/artifacts.rake]: minor: indentation

File: rakelib/artifacts.rake [hunk 5]
```diff
@@ -1,7 +1,18 @@
...
 
   SNAPSHOT_BUILD = ENV["RELEASE"] != "1"
-  PACKAGE_SUFFIX = SNAPSHOT_BUILD ? "-SNAPSHOT" : ""
+  VERSION_QUALIFIER = ENV["VERSION_QUALIFIER"]
+  if VERSION_QUALIFIER
+    if SNAPSHOT_BUILD
+      PACKAGE_SUFFIX = "-#{VERSION_QUALIFIER}-SNAPSHOT"
+    else
+      PACKAGE_SUFFIX = "-#{VERSION_QUALIFIER}"
+    end
```
Comment (colinsurprenant) [inline rakelib/artifacts.rake]: minor: this could be rewritten in 2 lines and makes that block more readable?

File: logstash-core-plugin-api/logstash-core-plugin-api.gemspec (hunk 3 above)
Comment (guyboertje) [reply logstash-core-plugin-api/logstash-core-plugin-api.gemspec]: No it is not, I only change it when I see gem build complain.

File: rakelib/artifacts.rake (hunk 5 above)
Comment (guyboertje) [reply rakelib/artifacts.rake]: ?? as a ternary operation ??

Comment (guyboertje) [reply rakelib/artifacts.rake]: NVM

File: build.gradle (hunk 2 above)
Comment (guyboertje) [reply build.gradle]: I tested the line in a Groovy online REPL. If there is no match then the `qualifiedVersion` string will be `"NULL"`.

If anything I should swap some lines around and use `qualifiedVersion` more.

If validation is necessary (not saying it is not), it should be done in the `task downloadEs(type: Download)` task.

Comment (guyboertje) [reply build.gradle]: I guess a doFirst closure that `throw new GradleException("could not find the current artifact from the artifact-api for version: ${version}")` if `qualifiedVersion` is "NULL"

**Comment by colinsurprenant** [comment]
overall LGTM - my only concern at this point is about the need of calling `artifacts-api.elastic.co`; is that necessary? and consequently the validation of the result from the api.

File: build.gradle (hunk 2 above)
Comment (guyboertje) [reply build.gradle]: In the groovy REPL `.toURL().text` throws an exception for an incorrect url `https://artifacts-api.elastic.co/v2/versions/`, so I guess that would halt the script in that case.
If the result is not a serialised JSON object then either the JSON parser throws an exception or the next line complains about a missing property on object `dlVersions` exception.

I think those are covered. I'll add the doFirst block.

Comment (guyboertje) [reply build.gradle]: Retested the full groovy block in the online repl incl. the `.toURL().text` with various "chaos-monkey" like inputs.

**Comment by colinsurprenant** [comment]
LGTM per discussion with @guyboertje and @jsvd
